projector-client-web/src/main/kotlin/org/jetbrains/projector/client/web/speculative/Typing.kt

```diff
@@ -145,27 +145,39 @@ sealed class Typing {
         rect(
           x = editorMetrics.x,
           y = editorMetrics.y,
-          w = editorMetrics.width - 10,  // todo: -10 to prevent scrollbar be repainted
+          w = editorMetrics.width,
           h = editorMetrics.height
         )
         clip()
 
-        val imageData = getImageData(
-          sx = firstCaretLocation.x,
-          sy = firstCaretLocation.y,
-          sw = editorMetrics.width - (firstCaretLocation.x - editorMetrics.x),
-          sh = currentCarets.nominalLineHeight.toDouble() + 2 + 2
-        )
-
-        putImageData(imageData, firstCaretLocation.x + currentCarets.plainSpaceWidth, firstCaretLocation.y)
-
         val fontFace = currentCarets.fontId?.toFontFaceName() ?: "Arial"
         val fontSize = "${currentCarets.fontSize}px"
 
         font = "$fontSize $fontFace"  // todo: use a proper font style
         fillStyle = "#888"  // todo: use a proper color
 
-        val stringYPos = firstCaretLocation.y + currentCarets.fontSize  // todo: offsetting by fontSize seems to be not exact
+        val speculativeCharWidth = measureText(event.char.toString()).width
+
+        val sw = editorMetrics.width - (firstCaretLocation.x - editorMetrics.x) - currentCarets.scrollBarWidth - speculativeCharWidth
+
+        if (sw > 0) { // check we actually have some graphics to move
+          val imageData = getImageData(
+            sx = firstCaretLocation.x,
+            sy = firstCaretLocation.y,
+            sw = sw,
```

Comment:
1. What does `sw` mean? Maybe a more clear name is needed.
2. The comment isn't needed much here because it's obvious.

I suggest changing to the following, feel free to correct me if I'm wrong:

```suggestion
        val widthToCopy = editorMetrics.width - (firstCaretLocation.x - editorMetrics.x) - currentCarets.scrollBarWidth - speculativeCharWidth

        if (widthToCopy > 0) {
          val imageData = getImageData(
            sx = firstCaretLocation.x,
            sy = firstCaretLocation.y,
            sw = widthToCopy,
```

Reply:
Renamed to widthToMove